### ISSUE TITLE
Docs: [no-constant-binary-expression] misleading info about newly constructed objects

### ISSUE THREAD

**Issue by kirkwaiblinger** [eslint/eslint#19048]
### Docs page(s)

https://eslint.org/docs/latest/rules/no-constant-binary-expression

### What documentation issue do you want to solve?

The doc page contains this passage 

> Additionally, this rule detects comparisons to newly constructed objects/arrays/functions/etc. In JavaScript, where objects are compared by reference, a newly constructed object can _never_ `===` any other value.

Technically, it's true. However, it's _not quite technically true_ that the result of a `new` invocation is never equal to another object, since a `new` invocation need not return a newly constructed object, as is somewhat implied by the fact that the rule reports on comparisons to `new` expressions.

See 
```js
const obj = {};
class Clazz {
    constructor () {
      return obj;
    }
}
new Clazz === obj; // true
new Clazz === new Clazz; // true;
```

maybe a more realistic example is 
```js
class Singleton {
    constructor() {
        Singleton.instance ??= this;
        return Singleton.instance;
    }
}

new Singleton === new Singleton
```

### What do you think is the correct solution?

Mention the `new` operator edge case either in the docs (where it would make sense to mention the Singleton pattern) or in the error report message. Currently the rule reports 

> Unexpected comparison to newly constructed object. These two values can never be equal.

Maybe instead?

"Unexpected comparison to newly constructed object. These two values can almost never be equal."

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
> Mention the `new` operator edge case either in the docs (where it would make sense to mention the Singleton pattern) or in the error report message. Currently the rule reports
> 
> > Unexpected comparison to newly constructed object. These two values can never be equal.
> 
> Maybe instead?
> 
> "Unexpected comparison to newly constructed object. These two values can almost never be equal."

Can you provide an example where this rule reports an error but the values can be equal? I remember this was discussed on the implementation PR and we eventually decided not to report on `new` for the exact reasons you described.

**Comment by mdjermanovic** [comment]
Here's the relevant code:

https://github.com/eslint/eslint/blob/d474443109762f3b92811df0411965cf64f595c2/lib/rules/no-constant-binary-expression.js#L351-L403

**Comment by kirkwaiblinger** [comment]
As a user I'd probably actually prefer if it _did_ report the `new MyClass === new MyClass` case, with manual disables for super-rare false positives. However that's just a niche suggestion, not a problem with how the rule works. Feel free to close!

**Comment by mdjermanovic** [comment]
Closing as it seems there's no need for a documentation update, but feel free to open a change request for the mentioned option.
